### ISSUE TITLE
Performance: large tool yaml files

### ISSUE THREAD

**Issue by erasche** [galaxyproject/ephemeris#139]
Opening this to start some discussion maybe over performance during installation of many tools (especially many which are already installed.)

Currently usegalaxy.eu's jenkins job which (weekly) ensures that everything in https://github.com/usegalaxy-eu/usegalaxy-eu-tools is installed to our server, takes around [3-5 hours](https://build.galaxyproject.eu/job/usegalaxy-eu/job/install-tools/buildTimeTrend) on average (less if it crashes due to #133, more if it finishes completely.)

What can we do to accelerate this?

- Could we fetch from galaxy a list of everything already installed + revisions? Surely that endpoint could be relatively quick since it'd be:
  ```
  select tool_shed, name, owner, changeset_revision from tool_shed_repository;
  ```

  And diff that locally before installing?
- Does anyone have any nice suggestions for diffing YAML files? Maybe I could find some ugly hack to find out which was the last built commit, and then diff the yamls and find tools + IDs, and just process those?

**Comment by mvdbeek** [comment]
I think there might be some unnecessary recursions when checking if a repo is already installed, if no new tools are to be added this should be fast.

**Comment by mvdbeek** [comment]
So I looked at https://build.galaxyproject.eu/job/usegalaxy-eu/job/install-tools/114/consoleFull, it seems the time spent per yaml file is more or less proportional to the number of repos installed at that step, making me think that there really isn't much that can be done. When I said there might be some unnecessary recursion I thought we'd be fetching the list of installed repos multiple times, but that's just once per yaml file.

**Comment by bgruening** [comment]
If we are not afraid of touching the TS code, the TS could provide a repository.yaml file, which lists all available repos, plus revisions etc ...


This will then be fetched and compared without yaml files and voila, no?

**Comment by mvdbeek** [comment]
I guess I don't understand what the issue is. To me it looks like the performance issue is that installing things in Galaxy is slow, no amount of yaml-munging will make that go faster ?

**Comment by bgruening** [comment]
Most of the time is spend to trigger the installation of a revision that is already installed I think. We could completely skip this and only call installation that is actually needed.

This will also speed up this step: https://build.galaxyproject.eu/job/usegalaxy-eu/job/update-trusted-tools/

But I guess @erasche plan is more effective. We could use https://pypi.org/project/structurediff/ to compare the Galaxy json with our tool yaml. But we do not have a nice API to get us that information easily, isn't it?

**Comment by hexylena** [comment]
Exactly, it isn't slow to install, it's slow to say "this thing is already installed, let's move on".

Hence my suggestion of "pull list of already installed things" to speed up the process.

**Comment by mvdbeek** [comment]
> Exactly, it isn't slow to install, it's slow to say "this thing is already installed, let's move on".

Well that's a bug. Do you have an example ?

**Comment by mvdbeek** [comment]
OK, I see that for fastqc the script tries to repeatedly install `e7b2202befea`. Can you see that revision on the Galaxy server ?

**Comment by bgruening** [comment]
Nope e7b2202befea is not installed as far as I can see.

**Comment by mvdbeek** [comment]
OK, so I think what's happening is that the galaxy install API doesn't allow updating repos without a version bump (the warning symbol). This is silly IMO, we should do that.

**Comment by rhpvorderman** [comment]
I don't understand entirely, because I refactored the shed-tools code some time ago to check for installed repos and skip them. https://github.com/galaxyproject/ephemeris/blob/2a57316fd31b3616f78eea22a3da54fe459bfe7b/src/ephemeris/shed_tools.py#L75

Checking **all** installed tools is only **1** API request. Shed-tools then only should try to install things which are not in the installed list.

@erasche Helena, it seems that your use case is installing a big tool yaml file on a server that already has a lot of tools installed. Is that the case? If so, I am wondering why shed-tools would not work correctly. The code to handle this use case is present.

**Comment by hexylena** [comment]
> big tool yaml file on a server that already has a lot of tools installed

Yes, we're syncing every week. 99.9% of the tools in the tool yaml are already installed, only 1-10 new tool revisions each time.

That sounds amazing that this should already work! That line you linked, looks like it's in 0.10.2, which should be what we have installed: https://github.com/usegalaxy-eu/usegalaxy-eu-tools/blob/master/requirements.txt#L2

That's really odd that it isn't working as expected.

**Comment by mvdbeek** [comment]
The core of the issue is what I said above, repos that were updated without a version bump are not installable to the new revision. So ephemeris will attempt to do that each time the changeset revision differs, which is always. That's not an ephemeris issue, it's a Galaxy issue.

**Comment by hexylena** [comment]
@mvdbeek  Should we move this ticket there then?

**Comment by rhpvorderman** [comment]
> The core of the issue is what I said above, repos that were updated without a version bump are not installable to the new revision. 

I am confused now.  So besides the changeset revision, galaxy also checks some sort of internal version number? I was under the (apparantly wrong) assumption that changeset revision should be the be-all end-all of tool versions in galaxy. 

Can you point to the code? Maybe constructing a workaround in ephemeris is not that hard. A list comprehension and an extra check and EU galaxy should be chugging along happily. (If we manage to extract the magic versions).

**Comment by mvdbeek** [comment]
@erasche that'd be good. @rhpvorderman yes, there are updates and installations. If the tool version isn't bumped we just do an hg update <ctx-rev> without creating a new ToolShedRepository in the database, instead updating the existing one. Among other things that's because you can't and shouldn't load more than one tool with the same tool id and version.

The API is silly though, who cares about the install process if you want to install a given changeset revision?

I don't think it's ephemeris job to figure out the installable vs updatable revisions (you could do that by checking for gaps in the installable revisions from the toolshed, then check whether any installed revision falls in the gap preceeding the requested revision). But that would really fix the wrong end of the problem IMO.
